modules/ftp.go

```diff
@@ -0,0 +1,189 @@
+package modules
+
+import (
+	"net"
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zmap/zgrab2"
+)
+
+// FTPScanResults is the output of the scan. Identical to the original from zgrab, with the addition of TLSLog.
+type FTPScanResults struct {
+	Banner      string         `json:"banner,omitempty"`
+	AuthTLSResp string         `json:"auth_tls_resp,omitempty"`
+	AuthSSLResp string         `json:"auth_ssl_resp,omitempty"`
+	TLSLog      *zgrab2.TLSLog `json:"tls,omitempty"`
+}
+
+// FTP-specific command-line flags. Taken from the original zgrab (TODO: should FTPAuthTLS be on by default?).
+type FTPFlags struct {
+	zgrab2.BaseFlags
+	zgrab2.TLSFlags
+	Verbose    bool `long:"verbose" description:"More verbose logging, include debug fields in the scan results"`
+	FTPAuthTLS bool `long:"authtls" description:"Collect FTPS certificates in addition to FTP banners"`
+}
+
+// FTPModule implements the zgrab2.Module interface
+type FTPModule struct {
+}
+
+// FTPScanner implements the zgrab2.Scanner interface, and holds the state for a single scan instance
+type FTPScanner struct {
+	config *FTPFlags
+}
+
+// FTPConnection holds the state for a single connection to the FTP server.
+type FTPConnection struct {
+	buffer  [1024]byte // temp buffer for sending commands -- so, never interleave sendCommand calls on a given connection
+	config  *FTPFlags
+	results FTPScanResults
+	conn    net.Conn
+}
+
+// ftp.init() registers the ftp zgrab2 module
+func init() {
+	var module FTPModule
+	_, err := zgrab2.AddCommand("ftp", "FTP", "Grab a FTP banner", 21, &module)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (m *FTPModule) NewFlags() interface{} {
+	return new(FTPFlags)
+}
+
+func (m *FTPModule) NewScanner() zgrab2.Scanner {
+	return new(FTPScanner)
+}
+
+func (f *FTPFlags) Validate(args []string) error {
+	return nil
+}
+
+func (f *FTPFlags) Help() string {
+	return ""
+}
+
+func (s *FTPScanner) Init(flags zgrab2.ScanFlags) error {
+	f, _ := flags.(*FTPFlags)
+	s.config = f
+	return nil
+}
+
+func (s *FTPScanner) InitPerSender(senderID int) error {
+	return nil
+}
+
+func (s *FTPScanner) GetName() string {
+	return s.config.Name
+}
+
+func (s *FTPScanner) GetPort() uint {
+	return s.config.Port
+}
+
+// ftpEndRegex matches zero or more lines followed by a numeric FTP status code and linebreak, e.g. "200 OK\r\n"
+var ftpEndRegex = regexp.MustCompile(`^(?:.*\r?\n)*([0-9]{3})( [^\r\n]*)?\r?\n$`)
+
+// FTPConnection.isOKResponse() returns true iff the given response code indicates success (e.g. 2XX)
+func (ftp *FTPConnection) isOKResponse(retCode string) bool {
+	// TODO: This is the current behavior; should it check that it isn't garbage that happens to start with 2 (e.g. it's only ASCII chars, the prefix is 2[0-9]+, etc)?
```

Comment:
Probably, but might probably good enough for now. I think that for the sake of moving forward, we should fix bugs in the code separately from moving them over unless they're horrendous, for the sake of forward progress.

Reply:
Works for me -- I just wanted to make a note so that I didn't lose track of it.

In general, is an in-source TODO sufficient for items like these, or should these go into github issues?

Reply:
I think for something like this a TODO comment is probably fine for now